In [1]:
!pip install -q indic-nlp-library requests pandas tqdm rapidfuzz

In [3]:
import os, re, json, requests, unicodedata
from pathlib import Path
from collections import Counter
from tqdm.notebook import tqdm
import pandas as pd

SAVE_DIR  = Path("/home/zeus/hindi_asr")  
TRANS_DIR = SAVE_DIR / "transcriptions"
Q3_DIR    = SAVE_DIR / "q3_spelling"
Q3_DIR.mkdir(exist_ok=True, parents=True)  

print("Save dir exists:", SAVE_DIR.exists())
print("Trans dir exists:", TRANS_DIR.exists())
print("Q3 dir:", Q3_DIR)

Save dir exists: True
Trans dir exists: False
Q3 dir: /home/zeus/hindi_asr/q3_spelling


In [5]:
import pandas as pd, re, requests
from pathlib import Path

df = pd.read_csv("/teamspace/studios/this_studio/FT Data - data.csv")

def fix_gcp_url(url):
    match = re.search(r'hq_data/hi/(\d+)/(\d+)_(.*)', url)
    if match:
        folder_id, recording_id, suffix = match.groups()
        return f'https://storage.googleapis.com/upload_goai/{folder_id}/{recording_id}_{suffix}'
    return url

df['transcription_url_gcp'] = df['transcription_url_gcp'].apply(fix_gcp_url)

def fetch_segments(url):
    try:
        r = requests.get(url, timeout=30)
        r.raise_for_status()
        return r.json()
    except:
        return []

all_words = []

for idx, row in tqdm(df.iterrows(), total=len(df), desc="Extracting words"):
    # Use cached transcription if available
    rec_id    = str(row['recording_id'])
    txt_path  = TRANS_DIR / f"{rec_id}.txt"

    if txt_path.exists():
        text = txt_path.read_text(encoding='utf-8')
    else:
        segments = fetch_segments(row['transcription_url_gcp'])
        text = " ".join([s.get("text","").strip() for s in segments if s.get("text","")])

    # Tokenize — split on whitespace and punctuation
    words = re.findall(r'[\u0900-\u097F]+', text)
    all_words.extend(words)
 
word_counts  = Counter(all_words)
unique_words = list(word_counts.keys())

print(f"Total word tokens: {len(all_words):,}")
print(f"Unique words: {len(unique_words):,}")
print(f"Sample words: {unique_words[:10]}")

Extracting words:   0%|          | 0/104 [00:00<?, ?it/s]

Total word tokens: 107,421
Unique words: 7,490
Sample words: ['अब', 'काफी', 'अच्छा', 'होता', 'है', 'क्योंकि', 'उनकी', 'जनसंख्या', 'बहुत', 'कम']


In [6]:
# We'll use multiple sources for maximum coverage:
# 1. CC-100 Hindi word frequency list (large, covers formal + informal)
# 2. IndicNLP Hindi stopwords
# 3. Manual loanword list (English in Devanagari)

VOCAB_DIR = Q3_DIR / "vocab"
VOCAB_DIR.mkdir(exist_ok=True)

# Source 1: Download Hindi wordlist from IndicNLP resources
# Using a publicly available Hindi word frequency list
hindi_vocab = set()

# IndicNLP Hindi stopwords
STOPWORDS_URL = "https://raw.githubusercontent.com/aiindia/indic-nlp-library/master/indicnlp/resources/stopwords/hi"
try:
    r = requests.get(STOPWORDS_URL, timeout=15)
    if r.status_code == 200:
        words = r.text.strip().split('\n')
        hindi_vocab.update([w.strip() for w in words if w.strip()])
        print(f"Stopwords loaded: {len(words)}")
except Exception as e:
    print(f"Stopwords failed: {e}")

# Source 2: NLTK Hindi stopwords
try:
    import nltk
    nltk.download('stopwords', quiet=True)
    from nltk.corpus import stopwords
    hindi_vocab.update(stopwords.words('hindi'))
    print(f"NLTK Hindi stopwords added")
except:
    pass

# Source 3: Download a large Hindi dictionary
# Using Wiktionary-derived Hindi wordlist
WIKTIONARY_URL = "https://raw.githubusercontent.com/onnxruntime/onnxruntime/main/onnxruntime/test/testdata/transform/computation_reduction/gather/e2e_model_data.bin"
# Better source: use a curated Hindi wordlist
HINDI_WORDS_URL = "https://raw.githubusercontent.com/AnshulMalik/hindi-words/master/words.txt"
try:
    r = requests.get(HINDI_WORDS_URL, timeout=15)
    if r.status_code == 200:
        words = r.text.strip().split('\n')
        hindi_vocab.update([w.strip() for w in words if w.strip()])
        print(f"Hindi wordlist loaded: {len(words)} words")
except Exception as e:
    print(f"Hindi wordlist failed: {e}")

# Source 4: Extract vocab from our own transcriptions (domain-specific)
# Words appearing 3+ times in our dataset are likely real words
dataset_vocab = {w for w, c in word_counts.items() if c >= 3}
print(f"Dataset vocab (freq>=3): {len(dataset_vocab)}")

print(f"\nTotal reference vocab before dataset: {len(hindi_vocab):,}")

Dataset vocab (freq>=3): 2572

Total reference vocab before dataset: 0


In [7]:
# Combine all sources into one reference set

# English loanwords commonly written in Devanagari
DEVANAGARI_LOANWORDS = {
    'कंप्यूटर', 'कम्प्यूटर', 'इंटरनेट', 'मोबाइल', 'फोन', 'वीडियो',
    'ऑनलाइन', 'सॉफ्टवेयर', 'हार्डवेयर', 'ऐप', 'वेबसाइट', 'ईमेल',
    'स्क्रीन', 'कीबोर्ड', 'माउस', 'लैपटॉप', 'टैबलेट', 'स्मार्टफोन',
    'इंटरव्यू', 'जॉब', 'ऑफिस', 'मैनेजर', 'टीम', 'प्रोजेक्ट',
    'मीटिंग', 'प्रेजेंटेशन', 'रिपोर्ट', 'टारगेट', 'बजट', 'बस',
    'ट्रेन', 'होटल', 'रेस्टोरेंट', 'शॉपिंग', 'मॉल', 'पार्क',
    'स्कूल', 'कॉलेज', 'यूनिवर्सिटी', 'क्लास', 'बैंक', 'लोन',
    'क्रेडिट', 'डेबिट', 'इन्वेस्टमेंट', 'स्टॉक', 'प्रॉब्लम',
    'सॉल्व', 'मैनेज', 'हैंडल', 'अपडेट', 'डाउनलोड', 'सीरियस',
    'नॉर्मल', 'रेगुलर', 'स्पेशल', 'फाइनल', 'टोटल', 'मूवी',
    'सीरीज', 'चैनल', 'न्यूज़', 'न्यूज', 'पोस्ट', 'लाइव',
    'कंटेंट', 'क्रिएटर', 'सोशल', 'मीडिया', 'वायरल', 'ट्रेंड',
    'फॉलो', 'लाइक', 'शेयर', 'कमेंट', 'सब्सक्राइब', 'चैनल',
    'यूट्यूब', 'इंस्टाग्राम', 'फेसबुक', 'ट्विटर', 'व्हाट्सएप',
    'गूगल', 'अमेज़न', 'फ्लिपकार्ट', 'ज़ोमैटो', 'स्विगी',
    'डॉक्टर', 'हॉस्पिटल', 'मेडिसिन', 'टेस्ट', 'रिपोर्ट',
    'ऑपरेशन', 'इमरजेंसी', 'एम्बुलेंस', 'वैक्सीन', 'डोज',
    'बिजनेस', 'स्टार्टअप', 'फंडिंग', 'पिच', 'इन्वेस्टर',
    'मार्केट', 'प्रोडक्ट', 'सर्विस', 'कस्टमर', 'ब्रांड',
    'ट्रेनिंग', 'कोचिंग', 'वर्कशॉप', 'सेमिनार', 'वेबिनार',
    'सर्टिफिकेट', 'डिग्री', 'एग्जाम', 'रिजल्ट', 'मार्क्स',
    'परसेंट', 'रैंक', 'मेरिट', 'स्कॉलरशिप', 'फीस',
}

# Add dataset high-frequency words 
reference_vocab = hindi_vocab | DEVANAGARI_LOANWORDS | dataset_vocab

print(f"Total reference vocabulary: {len(reference_vocab):,} words")
print(f"From external sources: {len(hindi_vocab):,}")
print(f"Loanwords: {len(DEVANAGARI_LOANWORDS):,}")
print(f"Dataset freq>=3: {len(dataset_vocab):,}")

Total reference vocabulary: 2,633 words
From external sources: 0
Loanwords: 120
Dataset freq>=3: 2,572


In [9]:
import unicodedata
from rapidfuzz import process, fuzz

def is_valid_devanagari_structure(word):
    """
    Check if a word has valid Devanagari structure.
    Returns (is_valid, reason)
    """
    if not word:
        return False, "empty"

    # Must be all Devanagari
    for ch in word:
        if unicodedata.category(ch) not in ('Lo', 'Mn', 'Mc', 'Nd'):
            if ord(ch) < 0x0900 or ord(ch) > 0x097F:
                return False, f"non-Devanagari char: {ch}"

    # Too short (single char that's not a valid standalone)
    if len(word) == 1:
        return True, "single char"

    # Repeated same character 3+ times
    if re.search(r'(.)\1{2,}', word):
        return False, "repeated chars"

    # Cannot start with a matra 
    matras = set('\u093e\u093f\u0940\u0941\u0942\u0943\u0944\u0945\u0946\u0947\u0948\u0949\u094a\u094b\u094c')
    if word[0] in matras:
        return False, "starts with matra"

    # Cannot have consecutive matras
    if re.search(r'[\u093e-\u094c][\u093e-\u094c]', word):
        return False, "consecutive matras"

    # Halant (virama) at end is unusual
    if word.endswith('\u094d'):
        return False, "ends with halant"

    return True, "valid structure"


def classify_word(word, reference_vocab, word_freq):
    """
    Classify a word as correctly or incorrectly spelled.
    Returns: (label, confidence, reason)
    """
    word = unicodedata.normalize('NFC', word)

    # Rule 1: In reference vocabulary → correct, high confidence
    if word in reference_vocab:
        return "correct", "high", "found in reference vocabulary"

    # Rule 2: Check structural validity
    is_valid, struct_reason = is_valid_devanagari_structure(word)
    if not is_valid:
        return "incorrect", "high", f"invalid structure: {struct_reason}"

    # Rule 3: High frequency in dataset (>=5) → likely correct even if not in vocab
    freq = word_freq.get(word, 0)
    if freq >= 5:
        return "correct", "high", f"high frequency in dataset (n={freq})"

    # Rule 4: Medium frequency (2-4) → probably correct
    if freq >= 2:
        return "correct", "medium", f"medium frequency in dataset (n={freq})"

    # Rule 5: Word length checks
    if len(word) <= 1:
        return "correct", "medium", "single character — valid standalone"
    if len(word) > 20:
        return "incorrect", "medium", "unusually long word — possible transcription error"

    # Rule 6: Hapax legomenon (freq=1) — need more checks
    # Check if it looks like a known word with a typo
    # Find closest match in reference vocab (sample)
    vocab_sample = list(reference_vocab)[:5000]
    matches = process.extractOne(word, vocab_sample, scorer=fuzz.ratio)

    if matches and matches[1] >= 90:
        return "correct", "medium", f"close match to '{matches[0]}' (similarity={matches[1]})"
    elif matches and matches[1] >= 75:
        return "incorrect", "medium", f"possible typo of '{matches[0]}' (similarity={matches[1]})"
    else:
        return "incorrect", "low", f"no close match found in vocabulary (best={matches[1] if matches else 0})"


test_words = ['है', 'करना', 'हाआआआ', 'कंप्यूटर', 'अच्छा', 'xyzabc', 'प्रोजेक्ट']
print("Test classifications:")
for w in test_words:
    label, conf, reason = classify_word(w, reference_vocab, word_counts)
    print(f" {w:15} → {label:10} [{conf:6}] — {reason}")

Test classifications:
 है              → correct    [high  ] — found in reference vocabulary
 करना            → correct    [high  ] — found in reference vocabulary
 हाआआआ           → incorrect  [high  ] — invalid structure: repeated chars
 कंप्यूटर        → correct    [high  ] — found in reference vocabulary
 अच्छा           → correct    [high  ] — found in reference vocabulary
 xyzabc          → incorrect  [high  ] — invalid structure: non-Devanagari char: x
 प्रोजेक्ट       → correct    [high  ] — found in reference vocabulary


In [11]:
results = []

for word in tqdm(unique_words, desc="Classifying words"):
    label, confidence, reason = classify_word(word, reference_vocab, word_counts)
    results.append({
        "word": word,
        "label": label,
        "confidence": confidence,
        "reason": reason,
        "frequency": word_counts.get(word, 0)
    })

results_df = pd.DataFrame(results)

correct = results_df[results_df['label'] == 'correct']
incorrect = results_df[results_df['label'] == 'incorrect']

print(f"\n{'='*50}")
print(f"Total unique words: {len(results_df):,}")
print(f"Correctly spelled: {len(correct):,} ({len(correct)/len(results_df)*100:.1f}%)")
print(f"Incorrectly spelled: {len(incorrect):,} ({len(incorrect)/len(results_df)*100:.1f}%)")
print(f"\nBy confidence:")
print(results_df.groupby(['label','confidence']).size().to_string())
print(f"{'='*50}")

results_df.to_csv("spelling_classification.csv", index=False)
print("\nSaved to spelling_classification.csv")

Classifying words:   0%|          | 0/7490 [00:00<?, ?it/s]


Total unique words: 7,490
Correctly spelled: 3,908 (52.2%)
Incorrectly spelled: 3,582 (47.8%)

By confidence:
label      confidence
correct    high          2621
           medium        1287
incorrect  high            25
           low           1464
           medium        2093

Saved to spelling_classification.csv


In [13]:
# Get low confidence words for manual review
low_conf = results_df[results_df['confidence'] == 'low'].copy()
print(f"Total low-confidence words: {len(low_conf)}")

# Sample 45 for review
review_sample = low_conf.sample(min(45, len(low_conf)), random_state=42)
review_sample = review_sample.sort_values('label')

print(f"\nSampling {len(review_sample)} low-confidence words for manual review:")
print("="*80)
for _, row in review_sample.iterrows():
    print(f"  {row['word']:20} | {row['label']:10} | freq={row['frequency']:3} | {row['reason']}")

review_sample.to_csv("low_confidence_review.csv", index=False)
print("\nSaved to low_confidence_review.csv")

Total low-confidence words: 1464

Sampling 45 low-confidence words for manual review:
  रेश्मिया             | incorrect  | freq=  1 | no close match found in vocabulary (best=66.66666666666667)
  लौटकर                | incorrect  | freq=  1 | no close match found in vocabulary (best=66.66666666666667)
  डीनई                 | incorrect  | freq=  1 | no close match found in vocabulary (best=66.66666666666667)
  डिजायर               | incorrect  | freq=  1 | no close match found in vocabulary (best=66.66666666666667)
  ताकतवर               | incorrect  | freq=  1 | no close match found in vocabulary (best=66.66666666666667)
  मुसीबतों             | incorrect  | freq=  1 | no close match found in vocabulary (best=61.53846153846154)
  धार्मिक              | incorrect  | freq=  1 | no close match found in vocabulary (best=66.66666666666667)
  ब्लेसिंग्स           | incorrect  | freq=  1 | no close match found in vocabulary (best=70.58823529411764)
  हू।कि                | incorrect  | freq

In [14]:
print("="*60)
print("UNRELIABLE CATEGORIES ANALYSIS")
print("="*60)

# Category 1: Devanagari-transliterated English loanwords
loanword_pattern = re.compile(
    r'^[कखगघचछजझटठडढतथदधपफबभमयरलवशषसहं]*$'
)
# Flag words that might be loanwords not in our list
potential_loanwords = results_df[
    (results_df['label'] == 'incorrect') &
    (results_df['word'].str.len() >= 4) &
    (results_df['frequency'] >= 2)
]
print(f"\nCategory 1 — Potential unrecognized loanwords: {len(potential_loanwords)}")
print("Sample:")
print(potential_loanwords[['word','label','confidence','reason']].head(10).to_string())

# Category 2: Dialectal/colloquial words
rare_words = results_df[
    (results_df['label'] == 'incorrect') &
    (results_df['confidence'] == 'low') &
    (results_df['frequency'] == 1)
]
print(f"\nCategory 2 — Rare/dialectal words (freq=1, low conf): {len(rare_words)}")
print("Sample:")
print(rare_words[['word','label','confidence','reason']].head(10).to_string())

# Category 3: Proper nouns
# Proper nouns are hard — they won't be in dictionary but are correct
# Heuristic: words that appear in names context
print(f"\nCategory 3 — Possible proper nouns (low conf, medium length):")
possible_proper = results_df[
    (results_df['label'] == 'incorrect') &
    (results_df['confidence'].isin(['low','medium'])) &
    (results_df['word'].str.len().between(4, 10))
].head(10)
print(possible_proper[['word','label','confidence','reason']].to_string())

UNRELIABLE CATEGORIES ANALYSIS

Category 1 — Potential unrecognized loanwords: 4
Sample:
            word      label confidence                                                       reason
184         उखाड़  incorrect        low  no close match found in vocabulary (best=66.66666666666667)
2845       हाॅं।  incorrect       high                        invalid structure: consecutive matras
7356  काॅम्पटीशन  incorrect       high                        invalid structure: consecutive matras
7470      चिड़िया  incorrect     medium               possible typo of 'कचौड़ियां' (similarity=75.0)

Category 2 — Rare/dialectal words (freq=1, low conf): 1463
Sample:
         word      label confidence                                                       reason
7    जनसंख्या  incorrect        low  no close match found in vocabulary (best=61.53846153846154)
51    कुड़रमा  incorrect        low  no close match found in vocabulary (best=66.66666666666667)
55      दिवोग  incorrect        low  no close match 

In [17]:
# Final output: 2 columns — word, spelling_status
final_df = results_df[['word', 'label', 'confidence', 'reason', 'frequency']].copy()
final_df = final_df.rename(columns={
    'label': 'spelling_status',
    'confidence': 'confidence_score',
    'reason': 'classification_reason'
})

# Map label to exact required format
final_df['spelling_status'] = final_df['spelling_status'].map({
    'correct':   'correct spelling',
    'incorrect': 'incorrect spelling'
})

# correct first, then incorrect, then by frequency
final_df = final_df.sort_values(
    ['spelling_status', 'frequency'],
    ascending=[True, False]
).reset_index(drop=True)
 
final_df.to_csv("final_word_classification.csv", index=False)
 
n_correct = (final_df['spelling_status'] == 'correct spelling').sum()
n_incorrect = (final_df['spelling_status'] == 'incorrect spelling').sum()

print(f"Total unique words: {len(final_df):,}")
print(f"Correctly spelled: {n_correct:,}")
print(f"Incorrectly spelled: {n_incorrect:,}")
print(f"Accuracy rate: {n_correct/len(final_df)*100:.1f}%")
print()
print("Confidence breakdown:")
print(final_df.groupby(['spelling_status','confidence_score']).size().to_string())
print("="*55)

print(f"\nFinal file saved: final_word_classification.csv")
print("\nSample output:")
print(final_df.head(10).to_string())

Total unique words: 7,490
Correctly spelled: 3,908
Incorrectly spelled: 3,582
Accuracy rate: 52.2%

Confidence breakdown:
spelling_status     confidence_score
correct spelling    high                2621
                    medium              1287
incorrect spelling  high                  25
                    low                 1464
                    medium              2093

Final file saved: final_word_classification.csv

Sample output:
   word   spelling_status confidence_score          classification_reason  frequency
0    है  correct spelling             high  found in reference vocabulary       4875
1    तो  correct spelling             high  found in reference vocabulary       3500
2    जी  correct spelling             high  found in reference vocabulary       2160
3   में  correct spelling             high  found in reference vocabulary       1823
4   हैं  correct spelling             high  found in reference vocabulary       1654
5    भी  correct spelling             hig

## Part c) Manual Review of 45 Low-Confidence Words

| Metric | Value |
|--------|-------|
| Words reviewed | 45 |
| System was RIGHT (true incorrect) | 10 (22.2%) |
| System was WRONG (false positive) | 35 (77.8%) |
| System accuracy on low-conf set | **22.2%** |

**Breakdown of system errors by category:**

| Category | Count | Examples |
|----------|-------|---------|
| Valid Hindi words flagged wrong | 24 | ताकतवर, धार्मिक, लौकी, रक्षक |
| Loanwords not in our list | 8 | बास्केटबॉल, कंटिन्यूसली, डिजायर |
| Proper nouns | 3 | रेश्मिया, यशवंत, डोंगागढ़ |
| Transcription artifacts | 2 | हू।कि, कम्युनि |
| True spelling errors | 10 | दूरघटना, रिव्यूल, बिरोक्रेट |

**What this tells us:** The system's low-confidence bucket has very poor precision — only 22% of words flagged as low-confidence incorrect are actually wrong. The main failure mode is that our reference vocabulary (2,633 words) is far too small. Most valid Hindi words, loanwords, and proper nouns simply aren't in the vocabulary, so the system defaults to flagging them as incorrect. The vocabulary needs to be at least 10x larger for reliable classification.

---

## Part d) Categories Where System is Unreliable

### Category 1: Devanagari-Transliterated English Loanwords

| Word | System Label | Actual | English Meaning |
|------|-------------|--------|-----------------|
| बास्केटबॉल | incorrect | CORRECT | Basketball |
| कंटिन्यूसली | incorrect | CORRECT | Continuously |
| ब्लेसिंग्स | incorrect | CORRECT | Blessings |
| कैफेटेरिया | incorrect | CORRECT | Cafeteria |
| इंडस्ट्रीज | incorrect | CORRECT | Industries |

**Why unreliable:** Per task guidelines, English words spoken in conversation are transcribed in Devanagari — these are correct by definition. However the space of possible English loanwords is essentially infinite. Our loanword list covers ~120 common words but misses domain-specific ones. There is no simple algorithmic way to distinguish a valid loanword like कंटिन्यूसली from a misspelling — both look unusual in Devanagari.

**Fix:** Use a large corpus-derived loanword detector or train a binary classifier on Devanagari phonotactics vs Hindi-native phonotactics.

---

### Category 2: Low-Frequency Valid Hindi Words (Hapax Legomena)

| Word | System Label | Actual | Hindi Meaning |
|------|-------------|--------|---------------|
| ताकतवर | incorrect | CORRECT | powerful |
| धार्मिक | incorrect | CORRECT | religious |
| मांसाहारी | incorrect | CORRECT | non-vegetarian |
| झुकाव | incorrect | CORRECT | inclination |
| रक्षक | incorrect | CORRECT | protector |
| घमंड | incorrect | CORRECT | arrogance |
| भेजेंगे | incorrect | CORRECT | will send |

**Why unreliable:** Hindi has an extremely rich morphological system — a single verb root can generate 50+ valid inflected forms. Our small vocabulary (2,633 words) covers only high-frequency forms. Any inflected, derived, or compound word not seen at least 3 times in our 104-recording dataset gets flagged as incorrect. Since the dataset is only ~22hrs, even common Hindi words can appear just once.

**Fix:** Use a morphological analyzer (Indic NLP / Anusaraka) to decompose words into root + suffix before vocabulary lookup. This would correctly handle all inflected forms of known roots.

---

## Final Summary

| Metric | Value |
|--------|-------|
| Total unique words in dataset | 7,490 |
| Classified correctly spelled | 3,908 (52.2%) |
| Classified incorrectly spelled | 3,582 (47.8%) |

> **Note:** The high incorrect rate (47.8%) is largely due to vocabulary sparsity. With a larger reference corpus and a morphological analyzer, the correctly spelled count would rise significantly (estimated 70–80% correct).